### PR TITLE
Caching layer: oracle scripts and revisions

### LINKED ISSUE #917: Implement caching layer for oracle script info
The oracle script table should contain
- ID (pk)
- name
- description
- owner
- request fee
- last update

Also add oracle script revision table that should contain
- ID (pk)
- Revision number (pk)
- Name
- timestamp
- block height
- tx hash

### PR DESCRIPTION
**Pull request by evilpeach**
Fixes #917

### CODE REVIEW

File: chain/db/oracle_script.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+	code []byte,
+	blockTime time.Time,
+) OracleScript {
+	return OracleScript{
+		ID:          id,
+		Name:        name,
+		Description: description,
+		Owner:       owner.String(),
+		Code:        code,
+		LastUpdated: blockTime.Unix(),
```
Comment (sorawit) [inline chain/db/oracle_script.go]: Please use millisecond.

File: chain/db/oracle_script.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+	)
+	err := b.tx.Create(&oracleScript).Error
+	if err != nil {
+		return err
+	}
+
+	return b.tx.Create(&OracleScriptRevision{
+		OracleScriptID: id,
+		Name:           name,
+		Timestamp:      blockTime.Unix(),
```
Comment (sorawit) [inline chain/db/oracle_script.go]: ditto

**Review by sorawit — CHANGES_REQUESTED** [review]
Please fix merge conflict.

File: chain/db/table.go [hunk 3]
```diff
@@ -47,6 +47,24 @@ type DataSourceRevision struct {
...
+	LastUpdated int64  `gorm:"not null"`
+}
+
+type OracleScriptRevision struct {
+	OracleScriptID int64  `gorm:"primary_key;auto_increment:false"`
+	RevisionNumber int64  `gorm:"primary_key"`
+	Name           string `gorm:"not null"`
+	Timestamp      int64  `gorm:"not null"`
+	BlockHeight    int64  `gorm:"not null"`
+	TxHash         []byte `gorm:"not null"`
```
Comment (sorawit) [inline chain/db/table.go]: Add fk to transaction

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/db/table.go (hunk 3 above)
Comment (sorawit) [reply chain/db/table.go]: @taobun should we add fk here? It will not work with oracle scripts generated during genesis......

Comment (taobun) [reply chain/db/table.go]: If we allow to have txHash 000000... as genesis state, it sounds make sense?

Comment (sorawit) [reply chain/db/table.go]: Yeah let's do that.

Comment (sorawit) [reply chain/db/table.go]: Three choices:
- Don't do FK
- Use dummy TX like you suggest
- Make `txHash` column nullable. NULL does not need to match FK.

Comment (taobun) [reply chain/db/table.go]: The answer is Make txHash column nullable. NULL does not need to match FK.

Comment (taobun) [reply chain/db/table.go]: watch #995 as example

**Review by taobun — APPROVED** [review]
LGTM

File: chain/db/table.go [hunk 4]
```diff
@@ -47,6 +47,24 @@ type DataSourceRevision struct {
...
+	Owner       string `gorm:"not null"`
+	Code        []byte `gorm:"not null"`
+	LastUpdated int64  `gorm:"not null"`
+}
+
+type OracleScriptRevision struct {
+	OracleScriptID int64  `gorm:"primary_key;auto_increment:false"`
+	RevisionNumber int64  `gorm:"primary_key"`
+	Name           string `gorm:"not null"`
+	Timestamp      int64  `gorm:"not null"`
```
Comment (sorawit) [inline chain/db/table.go:63]: We will need to remove this, but let's do it on a separate diff.

**Review by sorawit — APPROVED** [review]

**Pull request merged**